Feature: Device messages

bechols97 opened the pull request:
## Summary

This PR adds a feature for device messages (i.e. store function arguments on device to be handled by a host callback at a later time). The original idea for the feature is to have better error handling on the device with handling any output on with a host callback. This moves the code from Camp to RAJA due dependency on atomic operations.

## Design review

For the design, there are some open questions. (regarding these open questions please see the `Design notes` below)
1. Currently, this design supports a MPSC model with the expectation that the device side will produce messages while a single host thread consumes messages. 
2. Additionally, before consuming messages, the current implementation forces the stream to synchronize. 
3. If there are more messages than the buffer size, then those messages are lost. This is to avoid waits on the device side. However, would it be beneficial to have a configurable option to use a circular buffer?

### Design notes

Based on discussion from a meeting:
- As suggested in the meeting, there should be policies for the message queue to support different use cases and allows the message queue to be extended if there are any future use cases. 
- Based on current use cases, the default policy of the message queue should support a MPSC model, wait all should synchronize the stream, and not storing additional messages when the buffer is full. 
- The `message_handler` class that stores the callback should be move-only to prevent accidental copies to a lambda with supporting a view-like queue to copy to device kernels (as mentioned below by @MrBurmark). 

### Additional design notes

Based on further discussions from a meeting:
- [x] Ability to get range of messages (so that user can sort)
- [x] Ability to unsubscribe a callback
- [ ] ~~Ability to bind args for a callback, such as source location or strings.~~
  - If moving to C++20, `std::bind_front` can be used. However, if needed, this can be added in a future PR
- [ ] ~~Store number of missed messages~~
  - Posting messages returns a bool. Users can use this to get number of missed messages.
- [ ] ~~Remove duplicate messages~~
  - This would likely be a separate queue policy, which can be added in a future PR
- [ ] ~~Move interface to something similar to the Reducer interface, where kernel/forall loop gives correct resource.~~
  - For now, a `message_manager` will just have one resource. If need be in the future, this can be expanded to support multiple resources

bechols97 committed:
Moves message handling to RAJA from Camp

  - This moves message container to RAJA to avoid dependency
    issues with required atomic operation. Currently, testing and
    waiting for messages will block until the stream is synchronized.

bechols97 committed:
Adds unit tests for message handler

MrBurmark commented on test/unit/messages/test-messages.cpp:
```diff
@@ -0,0 +1,100 @@
+    test = val;   
+  });
+
+  ASSERT_EQ(msg.test_any(), false);
```
You might also assert that test is 0 here.

MrBurmark commented:
I think the design would benefit from a view like object that can be used in device kernels.

artv3 commented:
@bechols97 , can you add an example in the examples folder? I have a use case where this may be handy. In my application if I a thread gets a negative value, I  want to take note and output it at the end of the kernel by the root rank. Currently I am using printf and every thread that encounters the negative value spews information onto the screen.

bechols97 committed:
Adds additional assert's to tests

bechols97 committed:
Initial setup for queue view

  - This allows the message queue to be passed to RAJA kernels

bechols97 committed:
Updates to separate functionality

  - This also allows the message queue to be allocated with pinned
    memory when needed

bechols97 committed:
Adds example and some comments

  - Currently, example requires XNACK with HIP. (message queue should be using
    pinned memory so need to look into this)

rcarson3 commented:
@bechols97 so one of the libraries I maintain has a failure macro that on the host side throw an error with some useful error messages associated with it which can provide useful context for users / developers for why something failed. Do you think we could emulate something like with this framework if we could provide the absolute max size of the char array / string literal that we'd like and then at the error site have that passed into your message class?

MrBurmark commented:
I think it would be possible to add a fixed capacity string-like object that could be passed through this interface.

bechols97 commented:
@rcarson3 Yes, the original intention behind this idea is to be used as a device side error handler (though left to be more generic in case there are other use cases). As @MrBurmark mentioned, it would be possible to use a fixed-string object with the message handler. For string literals, you would want some type of fixed-string object / char array to store the string that is later passed to a host side callback. 

In addition to the fixed-string object, any type that is trivially destructible should work as well.

bechols97 committed:
Re-formats messaging code

bechols97 committed:
Fixes bug in example

MrBurmark commented:
The current state looks good if you're trying to handle a single kind of error.
```
auto logger = RAJA::message_handler<void(int*, int, int)>(num_messages, host, 
  [](int* ptr, int idx, int value) {
    std::cout << "\n pointer " << ptr << " a[" << idx << "] = " << value << "\n";
  }
);

auto cpu_msg_queue = logger.get_queue<RAJA::mpsc_queue>();
RAJA::forall<RAJA::seq_exec>(host, RAJA::RangeSegment(0, N), [=] (int i) { 
  if (a[i] < 0) { 
    cpu_msg_queue.try_post_message(a, i, a[i]); 
  }
});

logger.wait_all();
```

MrBurmark commented:
Another use case that I'm considering is having a long lived logger with an allocation. Then I can enqueue multiple types of messages in that while keeping the gpu running and check for messages occasionally to avoid extra synchronizes.
I am not sure this is feasible however as most of our logging use cases involve catching an error and stopping. If we did continue running we would likely encounter a hard error like a seg fault later.

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

bechols97 committed:
Adds factory functions and deduction guides for message_handler

bechols97 committed:
Runs clang-format

bechols97 committed:
Adds spsc queue policy

bechols97 committed:
Adds a fixed sized string to example

bechols97 committed:
Fixes formatting with clang-format

bechols97 committed:
Adds factory function to use default resource with execution policy

rcarson3 commented on examples/messages-forall.cpp:
```diff
@@ -0,0 +1,461 @@
+ *
+ */
+
+// This is a simplified example fixed string to show how
+// custom types can be used with the message queue.
+template <std::size_t N>
+class my_string
```
Thanks for adding an example with a "string-like" type in it :) It's helps provide me a road map for how I could replace my current Warning / Error reporting on the GPUs with this framework.

artv3 commented:
@bechols97 , can we bring this up to date?

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

gberg617 commented on include/RAJA/policy/msg_queue/spsc_queue.hpp:
```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#ifndef RAJA_mpsc_queue_HPP
```
```suggestion
#ifndef RAJA_spsc_queue_HPP
```

gberg617 replied:
Same for line 20 below and for line 6.

gberg617 commented on include/RAJA/policy/msg_queue/spsc_queue.hpp:
```diff
@@ -0,0 +1,109 @@
+ *
+ * \file
+ *
+ * \brief   Header file containing implementation for a MPSC
```
```suggestion
 * \brief   Header file containing implementation for a SPSC
```
Would it be worth defining what these acronyms (MPSC/SPSC) mean somewhere?

gberg617 commented on include/RAJA/policy/msg_queue/mpsc_queue.hpp:
```diff
@@ -0,0 +1,109 @@
+  {
+    if (m_container != nullptr)
+    {
+      auto local_size = RAJA::atomicInc<auto_atomic>(&(m_container->m_size));
```
Is the only difference between MPSC and SPSC the atomic add? I wonder if there's a way to combine these somehow to reduce code duplication.

MrBurmark replied:
I believe if you switch to the seq_atomic policy it will just do an add.

bechols97 replied:
Yes, the `seq_atomic` policy will just do an add. My original thought process was that the SPSC policies could have some additional optimizations with storing messages due to knowing that their aren't multiple publishers (though the default policy doesn't show a good example of this). However, I don't know if there would be any use cases for the SPSC policies. If there aren't any use cases at the moment for the SPSC, then these can be removed.

The MPSC policy was looking at using `auto_atomic` to support this with builds that have parallel sections with both GPUs and OpenMP in the same build.

MrBurmark replied:
SPSC will probably be the most used as its what the normal cpu only build would use.

bechols97 committed:
Refactors storing messages to be more generic

  - This now supports multiple message types with the same
    buffer. Currently, only supports basic queue.
  - Fixes some typos and adds to documentation

bechols97 committed:
Fixes build with example

bechols97 committed:
Fixes message alignment issues

bechols97 committed:
Formats code

rhornung67 committed:
Merge branch 'develop' into feature/bechols97/device_messages

MrBurmark commented on include/RAJA/policy/msg_queue/mpsc_queue.hpp:
```diff
@@ -0,0 +1,127 @@
+  using args_type = camp::tuple<Args...>;
+  using size_type = typename Container::size_type;
+
+  queue(int id, Container& container) : m_id {id}, m_container {&container} {}
```
It would be nice if the user didn't have to supply the id.

bechols97 replied:
While users can use this class directly, the preferred way would be for uses to keep the queue type returned when subscribing to a new message type. This will set the id automatically.

MrBurmark commented on include/RAJA/util/messages.hpp:
```diff
@@ -0,0 +1,332 @@
+    size_type m_begin {0};
+    size_type m_end {0};
+    size_type m_capacity {0};
+    size_type m_size {0};
```
Do you need size if you have begin and end?

bechols97 replied:
Size is now removed

MrBurmark commented on include/RAJA/util/messages.hpp:
```diff
@@ -0,0 +1,332 @@
+  using resource_type  = resource_deleter::resource_type;
+
+  message_bus()
+      : m_res {camp::resources::Host()},
```
Do we want a default constructor? If so it should probably delegate to the other constructor.

MrBurmark commented on include/RAJA/util/messages.hpp:
```diff
@@ -0,0 +1,332 @@
+      msg_args<msg_decay_t<Args>...>& aligned_args = *std::launder(
+          reinterpret_cast<msg_args<msg_decay_t<Args>...>*>(msg_args_buf));
+      camp::apply(c, aligned_args);
+      aligned_args.~msg_args<msg_decay_t<Args>...>();
```
Is the header destroyed somewhere?

bechols97 committed:
Changes to message callbacks

  - Adds wrapper class for message callbacks
  - Automatically gets message id

bechols97 committed:
Adds subscribe/unsubscribe functionality

bechols97 committed:
Adds tests for subscribe/unsubscribe member functions

bechols97 committed:
Fixes passing to callback list

bechols97 committed:
Removes extra size member variable

  - Also, fixes examples to not use stream 2 memory in stream 1

bechols97 committed:
Supports getting a container of all messages

  - This allows messages to be sorted, filtered, etc.

bechols97 committed:
Merge branch 'feature/bechols97/device_messages' of ssh://github.com/LLNL/RAJA into feature/bechols97/device_messages

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

bechols97 committed:
Formats code

MrBurmark commented on include/RAJA/policy/msg_queue/mpsc_queue.hpp:
```diff
@@ -0,0 +1,137 @@
+
+      // If message is not stored, update end to have the correct number of
+      // bytes and/or allow for additional smaller messages to be stored
+      RAJA::atomicSub<auto_atomic>(&(m_container->m_end), msg_sz);
```
There is a question here of if you ever want to allow end to be past capacity. There is the possibility of overflow to consider. There is the possibility that small messages will never actually see that there is unused capacity because of the ordering. It would be easier to reason about this if this used a CAS loop style atomic as then end would never go past capacity as updates would only ever be applied if the message would fit.

bechols97 replied:
This has been updated to use a CAS atomic loop for the updating the end size of the buffer.

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

bechols97 committed:
Fixes for cuda build and formatting

  - Moves internal `queue` class to be public
  - Moves typedef back to struct for message args

bechols97 committed:
Moves to CAS loop for atomic operation

bechols97 committed:
Fixes variable name

MrBurmark commented on examples/messages-forall.cpp:
```diff
@@ -0,0 +1,464 @@
+
+  std::cout << "\n Running RAJA simd_exec vector addition...\n";
+
+  RAJA::forall<RAJA::simd_exec>(host, RAJA::RangeSegment(0, N), [=] (int i) { 
```
Probably note that its not compatible with simd, if so.

bechols97 replied:
Since this feature is not compatible with simd, I just removed the loop from this example to avoid confusion.

MrBurmark commented on include/RAJA/pattern/messages/msg_header.hpp:
```diff
@@ -0,0 +1,52 @@
+  return ((size + alignment - 1) / alignment) * alignment;
+}
+
+struct msg_header
```
Instead of util you probably want to put the generic msg stuff in a directory in pattern.

bechols97 replied:
The msg headers have been moved to the pattern directory (the specific queue policies are still under the policy directory).

bechols97 committed:
Moves message headers to pattern folder and makes format

bechols97 committed:
Removes simd from example since not support with messages

bechols97 committed:
Destroys message header after use

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

bechols97 committed:
Fixes include path

bechols97 committed:
Fixes warnings

bechols97 committed:
Cleans up warnings

bechols97 committed:
Runs make style

artv3 commented on examples/messages-forall.cpp:
```diff
@@ -0,0 +1,452 @@
+
+  std::cout << "\n Running RAJA omp_parallel_for_static_exec (default chunksize) vector addition...\n";
+
+  RAJA::forall<RAJA::omp_parallel_for_static_exec< >>(host, RAJA::RangeSegment(0, N),
```
I think you may have extra `< >` in the policy?

artv3 replied:
oh I see, it has a default chunk size that you can modify: 
https://github.com/llnl/RAJA/blob/1d2ed09f2345497b41c4a13d8116c0244bddc38a/include/RAJA/policy/openmp/policy.hpp#L242-L244

bechols97 replied:
The chunk size doesn't really matter in this example, So, if you prefer having an explicit value in the `< >` to be more clear, then I don't mind updating this.

artv3 commented on examples/messages-forall.cpp:
```diff
@@ -0,0 +1,452 @@
+#elif defined(RAJA_ENABLE_SYCL)
+  RAJA::resources::Sycl res_gpu1;
+  RAJA::resources::Sycl res_gpu2;
+  using EXEC_POLICY = RAJA::sycl_exec<GPU_BLOCK_SIZE>;
```
I think you can simplify this by templating the resource on policy type. See: 
https://github.com/llnl/RAJA/blob/1d2ed09f2345497b41c4a13d8116c0244bddc38a/examples/forall-jit.cpp#L60

bechols97 replied:
In the purpose of this example, the hope was to show that this feature is able to work with non-default resources as well as work with multiple GPU resources. However, there are two examples showing the multiple non-default resources. If preferred, I can update the first example to only use the default stream (since most use cases will likely be using RAJA's default stream)?

artv3 commented on include/RAJA/pattern/messages/msg_header.hpp:
```diff
@@ -0,0 +1,55 @@
+namespace RAJA
+{
+RAJA_HOST_DEVICE
+constexpr std::size_t align(std::size_t size, std::size_t alignment = 16)
```
Should this go under utils? seems pretty useful beyond the MSG feature

bechols97 replied:
If preferred, I can move the `align` function to a separate header under the utils folder. Is there a particular header that this should belong in that currently exists under utils or just create a new header for it?

MrBurmark replied:
You might want to take a look around the utils dir and see if there is a header that this kind of fits with already.

rhornung67 replied:
@bechols97 please take a look in here https://github.com/llnl/RAJA/blob/develop/include/RAJA/util/align.hpp#L1

bechols97 committed:
Moves align to be under util directory and renames to align_sz

bechols97 committed:
Moves to use align_sz

bechols97 committed:
Adds missing header files

rhornung67 committed:
Merge branch 'develop' into feature/bechols97/device_messages

johnbowen42 commented on include/RAJA/pattern/messages/msg_callback.hpp:
```diff
@@ -0,0 +1,114 @@
+};
+
+template<typename Fn>
+struct get_signature;
```
We should move these into a generic `function_signature_helper.hpp` header.  See https://github.com/llnl/RAJA/pull/1949/changes#diff-72533564b1cbd49c320bfd7981489ca5e5a08143353955c70c3ef535e38fc4ccR56.  Arturo recently added similar methods for deducing the index of template parameters.  These types of metaprogramming utilities are broadly useful and should live in a more visible header

johnbowen42 replied:
It would be nice to move both to a generic header, maybe in `internal`.  I think it's good to consolidate stuff like this so we don't end up re-implementing the methods when we need it.  For example we have type_trait helper headers like https://github.com/llnl/RAJA/blob/develop/include/RAJA/pattern/kernel/TypeTraits.hpp

bechols97 replied:
Is there a preferred location for these types of metaprogramming utitlies? Would https://github.com/llnl/RAJA/tree/develop/include/RAJA/util be a good place for the more generic header?

johnbowen42 replied:
I think so yes, see for example https://github.com/llnl/RAJA/blob/develop/include/RAJA/util/EnableIf.hpp is already there.  I would try to move Arturo's helpers from [github.com/llnl/RAJA/pull/1949/changes#diff-72533564b1cbd49c320bfd7981489ca5e5a08143353955c70c3ef535e38fc4ccR56](https://github.com/llnl/RAJA/pull/1949/changes#diff-72533564b1cbd49c320bfd7981489ca5e5a08143353955c70c3ef535e38fc4ccR56.) there as well, and name it `FunctionSignatureUtil.hpp` or something

bechols97 replied:
This now updated and tried to de-duplicate similar code from Arturo's helpers.

johnbowen42 commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,422 @@
+  ~message_bus() { reset(); }
+
+  // Copy ctor/operator
+  message_bus(const message_bus&)            = delete;
```
I think we might also need `message_bus
(message_bus)            = delete;`

johnbowen42 commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,422 @@
+/// Currently, this forces a synchronize prior to calling
+/// the callback function or testing if there are any messages.
+///
+class message_manager
```
Do we want this case for the new classes @llnl/raja-core ?  `kernel` is all the `MessageManager` naming convention for classes

MrBurmark replied:
Generally we have been using CamelCase for classes.

bechols97 replied:
Updated to use CamelCase

johnbowen42 commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,422 @@
+  template<typename Callable>
+  void subscribe(msg_id id, Callable&& c)
+  {
+    auto callback = RAJA::msg_callback {std::forward<Callable>(c)};
```
it might be better to use an explicit constructor here

johnbowen42 commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,422 @@
+  {
+    auto& fn_list = m_callback_map.at(id);
+    auto it = std::find_if(fn_list.begin(), fn_list.end(), [](const auto& fn) {
+      return typeid(Callable).hash_code() == fn->hash();
```
I don't know if this is how we want to be hashing values.  For one, I think it's possible to have hashing collisions here, to `find_if` could just match the first of several possible matches. I think it might be better to hash together the `msg_id` with `std::type_index`, and just make `m_callback_map` a `std::unordered_map<HashCode, vector<msg_callback_t>>`

johnbowen42 commented on include/RAJA/policy/msg_queue/mpsc_queue.hpp:
```diff
@@ -0,0 +1,153 @@
+        size_type new_sz = old_sz + msg_sz;
+        local_sz         = old_sz;  // offset to start of message
+        // Checks if fits in queue
+        if (new_sz <= m_container->m_capacity)
```
do we ever resize the `m_container`?  this could be a race condition if so

bechols97 replied:
The `mpsc_queue` and `spsc_queue` are view-like containers to the owning version of `message_bus`. All resizing type of operations on `message_bus` will end up forcing the resource in `message_bus` to synchronize prior to resizing along with ignoring any messages that are currently stored.

artv3 commented on examples/messages-forall.cpp:
```diff
@@ -0,0 +1,452 @@
+#include "RAJA/util/resource.hpp"
+
+/*
+ *  Vector Addition Example
```
Maybe we can rename it to "RAJA::messages example" and update the description

bechols97 replied:
I updated the example to simplify and better showcase the different features. Let me know if anything else needs adjusting.

bechols97 committed:
Updates hashing for message callbacks

bechols97 committed:
Updates example to better describe and show different features

bechols97 committed:
De-duplicates function type traits

  - Both launch and messages were getting the first arg and the
    signature, which resulted in similar implementations. This
    tries to simplify both cases.

bechols97 committed:
Merge branch 'feature/bechols97/device_messages' of ssh://github.com/LLNL/RAJA into feature/bechols97/device_messages

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

bechols97 committed:
Fixes format

bechols97 committed:
Adds documentation for RAJA::messages

bechols97 committed:
Minor fixes to docs

bechols97 committed:
Expands on queue policies in documentation

bechols97 committed:
More edits to documentation

  - Simplifies example code
  - Renames msg_id to msg_queue_id

artv3 commented on docs/sphinx/user_guide/feature/messages.rst:
```diff
@@ -0,0 +1,167 @@
+passed to a function at a later time. For example, from a GPU, arguments can be stored
+and passed to function that prints to a file on the CPU. 
+
+.. warning:: This capability is new and should be considered experimental.
```
We have had feedback on how we advertise features as experimental and I would like share some thoughts: 
If this feature is meant for power users, I recommend we place under the expt namespace same if we expect API breaking changes in the near term. If this is pretty robust and we don't expect any API changes then maybe we don't need to advertise as experimental.  I share this because code projects have experienced development slow downs when updating to a RAJA version that introduces an API breaking change for a TPL. At any rate, maybe also expanding on considerations when integrating this feature into a project could mean. Even moving something from `RAJA::expt::` to standard `RAJA::` is a breaking change that users aren't a fan of.

artv3 commented on examples/messages-forall.cpp:
```diff
@@ -0,0 +1,393 @@
+int main(int RAJA_UNUSED_ARG(argc), char **RAJA_UNUSED_ARG(argv[]))
+{
+
+  std::cout << "\n\nRAJA vector addition example...\n";
```
```suggestion
  std::cout << "\n\nRAJA RAJA::messages with vector addition example...\n";
```

artv3 commented on examples/messages-forall.cpp:
```diff
@@ -0,0 +1,393 @@
+  msg_manager.unsubscribe(cpu_msg_queue.get_id(), err_callback);
+// _raja_msg_unsubscribe_end
+  
+  const int N = 100000;
```
```suggestion
  constexpr int N = 100000;
```

bechols97 committed:
Apply suggestions to example from code review

Co-authored-by: Arturo Vargas <artvar3@yahoo.com>

artv3 commented on examples/messages-forall.cpp:
```diff
@@ -0,0 +1,393 @@
+
+  host.deallocate(a_);
+  host.deallocate(b_);
+  host.deallocate(c_);
```
minor nitpicking, I think we should call this `res_host`  for consistency with the above example. It also helps give people give context.

bechols97 committed:
Updates host resource variable name to be more consistent

bechols97 committed:
Adds some considerations for applications

artv3 commented on test/unit/messages/test-messages.cpp:
```diff
@@ -0,0 +1,303 @@
+}
+
+TEST(message_handler, wait_all_overflow) {
+  // TODO: implement 
```
Will you add this in this PR?

bechols97 replied:
These were originally test cases for another queue policy. For the purposes of this PR, this won't be added back. If desired later, these extra policies can be added in a future PR without an API change. I will remove these.

artv3 replied:
Sounds good, thanks. Maybe future work can be captured in a GitHub issue

artv3 commented on test/unit/messages/test-messages.cpp:
```diff
@@ -0,0 +1,303 @@
+} 
+
+TEST(message_handler, try_post_message_overwrite) {
+  // TODO: implement
```
Another here

artv3 commented on include/RAJA/policy/msg_queue/mpsc_queue.hpp:
```diff
@@ -0,0 +1,153 @@
+
+// TODO: turning off for now
+// need to relook at logic
+#if 0
```
I saw this in a couple of places, it might be worth also documenting why this didn't get used and how you intend to use it. Also adding a date.

bechols97 committed:
Removes unused test

bechols97 committed:
Updates docs

bechols97 committed:
Updates comments on incomplete policies

artv3 reviewed (approved):
Just need to get the CI passing, but super excited for this contribution!

bechols97 committed:
Formats code

rhornung67 commented on docs/sphinx/user_guide/feature/messages.rst:
```diff
@@ -0,0 +1,184 @@
+Messages
+===============================
+
+RAJA provides a portable interface and type-safe way to store function arguments that
```
```suggestion
RAJA provides a portable interface and type-safe way to store function arguments that are
```

rhornung67 commented on docs/sphinx/user_guide/feature/messages.rst:
```diff
@@ -0,0 +1,184 @@
+passed to a function at a later time. For example, from a GPU, arguments can be stored
+and passed to function that prints to a file on the CPU. 
+
+.. warning:: This capability is new and should be considered experimental.
```
Since we decided to not put it in the experimental namespace, I would suggest changing this comment to something along the lines of `This capability is new. We would like users to try it out and give us feedback to improve it.`

rhornung67 commented on docs/sphinx/user_guide/feature/messages.rst:
```diff
@@ -0,0 +1,184 @@
+    :language: C++
+
+``buf_sz`` is the size of the buffer that stores messages. ``res_host`` is the resource of the 
+execution policy, which determines what type of memory to store messages with. For GPU resources,
```
```suggestion
execution policy, which determines the memory space in which messages will be stored. For GPU resources, for example,
```

rhornung67 commented on docs/sphinx/user_guide/feature/messages.rst:
```diff
@@ -0,0 +1,184 @@
+
+Handling messages
+^^^^^^^^^^^^^^^^^
+Lastly, there needs to be a way to "handle" messages. In others, direct the messages to the corresponding callback(s).
```
```suggestion
Lastly, there needs to be a way to direct messages to the corresponding callback(s).
```
I am confused by this. Is this what you mean?

bechols97 replied:
Yes, that was what I meant. I had some typos and missing words in there that made it confusing.

rhornung67 commented on docs/sphinx/user_guide/feature/messages.rst:
```diff
@@ -0,0 +1,184 @@
+Application considerations
+--------------------------
+
+Things to consider when using ``RAJA::messages`` in an application.
```
```suggestion
There are several things to consider when using ``RAJA::messages`` in an application.
```

rhornung67 commented on examples/messages-forall.cpp:
```diff
@@ -0,0 +1,390 @@
+ *  - to print to a file on the GPU using the RAJA::messages
+ *  - to interact with multiple GPU streams
+ *  - to create custom types that can be stored (note: these not to be
+ *    trivially destructible and copyable.
```
```suggestion
 *  - to create custom types that can be stored (note: these not to be
 *    trivially destructible and copyable)
```
I don't understand `these not to be`. What do you mean

bechols97 replied:
Looks like typo in the comment. I meant to say that "these types should be trivially destructible and trivially copyable". This should now be updated.

rhornung67 commented on examples/messages-forall.cpp:
```diff
@@ -0,0 +1,390 @@
+ *
+ */
+
+// This is a simplified example fixed string to show how
```
```suggestion
// This is a simplified example fixed length string to show how
```

bechols97 committed:
Apply suggestions to documentation from code review

Co-authored-by: Rich Hornung <hornung1@llnl.gov>

rhornung67 reviewed (changes requested):
Seems like a solid first draft. 

I have several small requests for documentation changes.

bechols97 committed:
More documentation fixes based on feedback

bechols97 committed:
Update comment in examples/messages-forall.cpp based on suggestion in PR

Co-authored-by: Rich Hornung <hornung1@llnl.gov>

johnbowen42 commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,422 @@
+  }
+
+  msg_bus m_bus;
+  std::vector<msg_fn_list_t> m_callback_map;
```
@bechols97 and I discussed offline that we should replace this with either (a) `std::unordered_map<int, std::unordered_map<decltype(std::hash<Callable>), std::vector<msg_callback_t>>>` or (b) implement a hash combiner in a common header and `std::unordered_map<CombinedHashT, std::vector<msg_callback_t>>`

johnbowen42 replied:
Even though `msg_id` is dense initially (assigned in ascending order 1,2,3), I think that if a single manager is used across many kernel launches this could lead to a memory leak, and we should probably `erase` the `msg_id` storage entry after its used

bechols97 replied:
Updated to something similar to option (a) with the message header storing an additional `std::size_t` for the hash of the `Callable`.

Also, yes that style (assuming there is a new message type for each kernel launch) would slowly consume more memory and would essentially be a memory leak. I updated the `unsubscribe_all(msg_id)` to `erase` the entry instead of just removing the callback list. However, this still requires users/applications to call either `unsubscribe_all(msg_id)` or `unsubscribe_all()` directly to remove an entry or all entries. This could likely be addressed in a future PR with an additional queue policy that deletes the entry once the queue goes out of scope.

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

bechols97 committed:
Updates mapping of messages to callbacks to use a map instead of vector

bechols97 committed:
Updates comments

bechols97 committed:
Updates unsubscribe_all(msg_id) to erase entry

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

rhornung67 committed:
Merge branch 'develop' into feature/bechols97/device_messages

bechols97 committed:
Adds optional allocator

  - The default uses a resource allocator with pinned memory

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

bechols97 committed:
Fixes build

bechols97 committed:
Adds back placement new for creating object lifetime

bechols97 committed:
Adds missing typename

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

MrBurmark commented on docs/sphinx/user_guide/feature/messages.rst:
```diff
@@ -0,0 +1,184 @@
+ mpsc                    Supports multiple producers,
+                         a single consumer; i.e., 
+                         requires atomic operations. 
+                         Automiatically determines 
```
```suggestion
                         Automatically determines 
```

bechols97 committed:
Update docs/sphinx/user_guide/feature/messages.rst

Co-authored-by: Jason Burmark <MrBurmark@users.noreply.github.com>

MrBurmark commented on examples/messages-forall.cpp:
```diff
@@ -0,0 +1,392 @@
+ *  The purpose of this example to show how RAJA::messages can be used.
+ *  For this example, the kernels will be launched using RAJA::forall 
+ *  along with a vector addition calculation. This will show how messages
+ *  can be used:
```
```suggestion
 *  The purpose of this example to show how RAJA::messages can be used.
 *  For this example, the kernels will be launched using RAJA::forall 
 *  along with a vector addition calculation. Messages can be used:
```
Would something like this make more sense?

bechols97 replied:
I updated this description to make more sense

MrBurmark commented on include/RAJA/pattern/messages/msg_callback.hpp:
```diff
@@ -0,0 +1,86 @@
+
+  template<typename Object>
+  explicit msg_callback(Object&& obj) : m_callable {std::move(obj)}
+  {}
```
Did you mean to use Callable, or did you want to template on Object?

bechols97 replied:
Yeah, I meant Callable here. I will update this, thanks

MrBurmark commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,444 @@
+      : m_res {res},
+        m_alloc {alloc},
+        m_bus {new(queue_allocator(alloc).allocate(1)) queue {},
+               resource_deleter {m_res, alloc}}
```
The deleter also gets a copy of the resource and allocator?

bechols97 replied:
The resource is needed to synchronize before the deleter can deallocate the memory buffer for the queue, and a custom allocator from a user is not guaranteed to synchronize.

MrBurmark replied:
I also don't see anything about checking whether or not the allocator supports propagation on copy. See https://en.cppreference.com/cpp/named_req/Allocator where it talks about things like what equality means for allocators and how things propagate.

MrBurmark replied:
Sure, but if those exist in the resource deleter does it make sense to use those instead of having two copies?

bechols97 committed:
Updates to use camel case and fixes typename

bechols97 committed:
Adjusts documentation

bechols97 committed:
Fixes comparison operator

bechols97 committed:
Moves to one allocator and resource

  - Using getters when needed

bechols97 committed:
Fixes move assignment for propagation on move assignment

bechols97 committed:
Updates documentation for custom allocator

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

MrBurmark commented on include/RAJA/pattern/messages/msg_callback.hpp:
```diff
@@ -0,0 +1,83 @@
+  explicit MsgCallback(Callable&& callable) : m_callable {std::move(callable)}
+  {}
+
+  std::type_index get_type() const final override { return typeid(Callable); }
```
final implies override
```suggestion
  std::type_index get_type() const final { return typeid(Callable); }
```

MrBurmark commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,483 @@
+    {
+      return !(*this == other);
+    }
+
```
Not needed in c++20
```suggestion
```

MrBurmark commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,483 @@
+  template<typename Resource>
+  MessageBus(Resource res, Allocator alloc = Allocator {})
+      : m_bus {new(queue_allocator(alloc).allocate(1)) Queue {},
+               ResourceDeleter {res, alloc}}
```
Its best to copy/move the allocator, then use the version that's being stored long term to allocate the bus.

bechols97 replied:
Good point, this has been updated

MrBurmark commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,483 @@
+  {
+    RAJA::MsgCallback callback {std::forward<Callable>(c)};
+    auto& fn_list = m_callback_map.at(id);
+    auto it = std::find_if(fn_list.begin(), fn_list.end(), [](const auto& fn) {
```
ranges version?

bechols97 replied:
Updated to use ranges version

MrBurmark commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,483 @@
+    {
+      for (const auto& msg : messages)
+      {
+        msg_id id = std::make_pair(msg->type, msg->hash);
```
Is this a message id or a callback id?

bechols97 replied:
This is a message id. The mapping structure is message id -> callback list. To provide a unique hash, a pair of `type_index(Callable)` and the old incrementing hash is created.

MrBurmark commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,483 @@
+    clear();
+  }
+
+  void wait_all()
```
Should this call handle_all?

bechols97 replied:
At one point `wait_all` and `handle_all` did slightly different things. However, yes, this has been updated to use handle_all to deduplicate code.

MrBurmark commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,483 @@
+
+  auto get_messages()
+  {
+    std::vector<const MsgHeader*> messages;
```
Could this generate the values on the fly instead of storing them?

bechols97 replied:
The purpose of `get_messages` storing messages was to return something that could:
  - Apply some operation to the message list, such as sorting
  - Avoid object lifetime issues when iterating through messages while destroying the headers

It is possible to generate messages on the fly and avoid object lifetime issues with having more specific loops in `wait_all`. However, I couldn't think of a good way to support sorting though I am open to suggestions.

MrBurmark replied:
Ok, we'll think about it more if needed.

MrBurmark commented on include/RAJA/policy/msg_queue/mpsc_queue.hpp:
```diff
@@ -0,0 +1,117 @@
+namespace messages
+{
+
+template<typename Container, typename... Args>
```
Could this be more descriptive than Container?

bechols97 replied:
Updated the name to be more descriptive than Container. Let me know if the new name works better.

MrBurmark commented on include/RAJA/policy/msg_queue/mpsc_queue.hpp:
```diff
@@ -0,0 +1,117 @@
+        m_container {&container}
+  {}
+
+  Queue(std::pair<std::size_t, std::size_t> id, Container* container)
```
Can container be nullptr?

bechols97 replied:
At the moment, if the container is a `nullptr` when the queue is created, then the message bus has gotten into a bad state.

MrBurmark commented on include/RAJA/policy/msg_queue/mpsc_queue.hpp:
```diff
@@ -0,0 +1,117 @@
+        // Checks if fits in queue
+        if (new_sz <= m_container->m_capacity)
+        {
+          old_sz = RAJA::atomicCAS<auto_atomic>(&(m_container->m_end), local_sz,
```
We are in the internals of RAJA, woud it be more clean to use the internal CAS loop implementation that takes lambdas?

MrBurmark commented on include/RAJA/policy/msg_queue/mpsc_queue.hpp:
```diff
@@ -0,0 +1,117 @@
+          local_sz + msg_sz <= m_container->m_capacity)
+      {
+        char* buf = m_container->m_data + local_sz;
+        new (buf) MsgHeader {args_sz, m_type, m_hash, buf + header_sz};
```
Does it make more sense to store the whole id type in the header?

bechols97 replied:
If I understand this comment correctly, this is asking if storing m_type and m_hash as a std::pair in the message header as well would make more sense, correct? The reason I didn't do this is because it require `constexpr` functions to used in `__host__ __device__` functions. This is possible since `std::pair` is `constexpr` but would force users to use `--expt-relaxed-constexpr` with NVCC, which I don't believe is a requirement yet.

MrBurmark replied:
Oh, right, annoying.

MrBurmark commented on include/RAJA/util/HashCombiner.hpp:
```diff
@@ -0,0 +1,68 @@
+ *
+ * Note: this is a simple hash combiner that can generate
+ * collisons with being useful for combining hashes for
+ * maps and sets.
```
Is this all you really meant to say?
```suggestion
 * Note: this is a simple hash combiner that can generate
 * collisons.
```

MrBurmark replied:
Do we want to make a generic util given the note above?

bechols97 replied:
Should I move this to a `detail` or internal namespace then?

MrBurmark replied:
Ya anything that is an implementation detail and shouldn't be called by users should be in a namespace like detail.

MrBurmark commented on include/RAJA/util/ResourceAllocator.hpp:
```diff
@@ -0,0 +1,118 @@
+namespace detail
+{
+template<typename Resource>
+struct ResourceAllocator
```
Should camp provide this class.

bechols97 replied:
For traceability, the PR in camp for this functionality here: https://github.com/llnl/camp/pull/207. Once this is merged, I can update the camp submodule and use the camp version instead.

bechols97 replied:
Updated this to use the camp `ResourceAllocator`

MrBurmark commented on include/RAJA/util/ResourceAllocator.hpp:
```diff
@@ -0,0 +1,118 @@
+    {
+      return !(lhs == rhs);
+    }
+
```
Handled by c++20.
```suggestion
```

MrBurmark commented on include/RAJA/util/ResourceAllocator.hpp:
```diff
@@ -0,0 +1,118 @@
+      m_res.deallocate(ptr, m_mem_type);
+    }
+
+    Resource const& get_resource() const { return m_res; }
```
How about a non-const accessor as well?

MrBurmark commented on test/unit/CMakeLists.txt:
```diff
@@ -19,3 +19,4 @@ add_subdirectory(view-layout)
 add_subdirectory(algorithm)
 add_subdirectory(workgroup)
 add_subdirectory(indexing)
+add_subdirectory(messages)
```
How about functional tests with forall to ensure that this works correctly with each backend?

bechols97 replied:
Add some functional tests

bechols97 committed:
Apply suggestions from code review

Co-authored-by: Jason Burmark <MrBurmark@users.noreply.github.com>

bechols97 committed:
Adds ranges version of find_if

bechols97 committed:
Reduces duplicate code

bechols97 committed:
Adds non-const accessors

bechols97 committed:
Updates CAS loop to use generic atomic in RAJA

  - Gives more descriptive name for container type
  - Makes explicit ctor

bechols97 committed:
Adjusts to use long stored allocator

bechols97 committed:
Moves hash_combine to detail namespace

bechols97 committed:
Adds some functional tests for messages

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

artv3 commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,479 @@
+  {
+    using value_type     = char;
+    using size_type      = unsigned long long;
+    using pointer        = value_type*;
```
what is the rule of thumb for ending the alias with _type?

bechols97 replied:
These are just following the naming convention from the C++ standard library.

artv3 replied:
ah okay thanks!

artv3 commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,479 @@
+    using msg_callback_t = decltype(callback);
+    if (it != fn_list.end())
+    {
+      // TODO: would it be better to throw or just replace old one?
```
Do we want to sort this out before merging? or an update?

bechols97 replied:
Personally, I am in favor of just replacing old callbacks, which is the current behavior. But, I am open to changing this if there are other opinions.

artv3 replied:
I'm good with that, I suggest removing the TODO if we stick in that direction

bechols97 replied:
Removed the TODO comment

bechols97 committed:
Removes TODO comment

rhornung67 committed:
Merge branch 'develop' into feature/bechols97/device_messages

bechols97 committed:
Updates CAMP submodule

bechols97 committed:
Fixes build after camp update

bechols97 committed:
Uses camp's resource allocator

bechols97 committed:
Fixes build

bechols97 committed:
Attempts to fix SYCL build

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

bechols97 committed:
Fixes syntax error with SYCL

bechols97 committed:
Merge branch 'feature/bechols97/device_messages' of ssh://github.com/LLNL/RAJA into feature/bechols97/device_messages

bechols97 committed:
Fixes build error with SYCL

bechols97 committed:
Updates for SYCL build

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

artv3 commented on docs/sphinx/user_guide/feature/messages.rst:
```diff
@@ -0,0 +1,195 @@
+
+===============================
+Messages
+===============================
```
It might help to be more direct in the intro here, the opening paragraph mentions messages at the end but I think it could be better to start with the headlining capability.

bechols97 replied:
I updated the intro a bit to provide better context for the capability. Please feel free to let me know if the intro or other sections need to be updated.

bechols97 committed:
Minor updates to intro of documentation

bechols97 committed:
Backs of cmake version due to json string parsing issue in cmake/4.4

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

bechols97 reviewed (commented):
@lc-hubcast approve

bechols97 reviewed (commented):
@lc-hubcast approve

bechols97 reviewed (commented):
@lc-hubcast approve

rhornung67 committed:
Merge branch 'develop' into feature/bechols97/device_messages

bechols97 committed:
Minor fixes:
  - Fixes bug with message manager allocator defaults
  - Adds concepts to factory function

bechols97 reviewed (commented):
@lc-hubcast approve

bechols97 committed:
Fixes memory type in example

bechols97 committed:
Fixes lifetime of messages

bechols97 committed:
Fixes move operator

bechols97 committed:
Adds extra check to verify callbacks have same argument types

bechols97 committed:
Moves to use counter that always increments instead of size that could be duplicated

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

bechols97 committed:
Removes temporary solution for cmake issue on mac

bechols97 committed:
Uses DATA_ALIGN for alignment size in messages

bechols97 committed:
Fixes formatting and minor updates
 - Updates unsubscribe_all(id) to allow users to re-subscribe to
   the message id
 - Adds an erase_all(id) to erase the entry from the callback map

bechols97 committed:
Fixes test

bechols97 committed:
Empty commit to restart CI

bechols97 committed:
Merge branch 'develop' into feature/bechols97/device_messages

MrBurmark commented on include/RAJA/pattern/messages/msg_manager.hpp:
```diff
@@ -0,0 +1,521 @@
+template<concepts::Resource Resource>
+auto make_message_manager(std::size_t bus_sz, Resource r)
+{
+  using Allocator = RAJA::ResourceAllocator<char, Resource>;
```
I'm not sure we want to default this behavior if we want people to use something like umpire instead.

bechols97 replied:
Hmm, maybe I am missing something and can certainly remove the default if preferred. But from our earlier discussion, I thought the reason for adding the `ResourceAllocator` was to have a default allocator with users still having an option to use a more specialized allocator (like an umpire allocator) if need be?

bechols97 replied:
Removed default allocator

bechols97 committed:
Removes factory function for default allocator

bechols97 committed:
Updates the default alignment value

bechols97 committed:
Adjusts documentation on allocators used in the message manager

bechols97 committed:
More adjustments to allocator section

MrBurmark commented on docs/sphinx/user_guide/feature/messages.rst:
```diff
@@ -0,0 +1,201 @@
+requirements for an allocator and will be used to allocate/deallocate
+the message bus. The message bus will always read messages on the host
+to forward the message to the respect callback(s). However, depending on the build
+and where messages are store, messages may be written on either the host
```
```suggestion
and where messages are stored, messages may be written on either the host
```

bechols97 committed:
Update docs/sphinx/user_guide/feature/messages.rst

Co-authored-by: Jason Burmark <MrBurmark@users.noreply.github.com>

rhornung67 committed:
Merge branch 'develop' into feature/bechols97/device_messages
